### ISSUE TITLE
Alternatives to NSIS

### ISSUE THREAD

**Issue by jaimergp** [conda/constructor#646]
I think it's safe to say NSIS is not the most development friendly engine ever, specially coming over from the Python world. While incredibly performant, powerful, customizable and extensible, it comes with a few gotchas:

- Tricky to debug and test. The hoops we go through in the `run_examples.py` showcase this nicely.
- Difficult-to-obtain logs. No output in stdout!
- Scripting syntax is sometimes too low-level.

Ideally, I would love to see if we can find a replacement with the following features:

- Free and redistributable (and open-source?)
- Scriptable in a sane way that allows us to run custom code
- Customizable and extensible UI (branding, text, extra pages)
- Good development/debugging experience (CLI installs should be as informative as GUI ones)

[This website](http://www.installsite.org/pages/en/msi/authoring.htm) ([plus this one](http://www.installsite.org/pages/en/tt_nonmsi.htm)) has TONS of info about the whole landscape. I'll summarize the main findings here, focusing on free products:

- [MSI installers](http://www.installsite.org/pages/en/w2k_aboutmsi.htm)
  - The "canonical" solution (like PKGs in macOS).
  - Robust, reliable, but not extensible.
  - It even has a standard library package, [`msilib`](https://docs.python.org/3/library/msilib.html) (deprecated in 3.11 though). Low-level.
- [WiX](https://wixtoolset.org/)
  - An abstraction on top of the MSI technology
  - Used by other tools like `briefcase` ([see code](https://github.com/beeware/briefcase/blob/8d503064b40313e2a0c4559e59cdc7c24b72adac/src/briefcase/platforms/windows/__init__.py#L107))
  - [Customizable](https://wixtoolset.org/docs/v3/wixui/wixui_customizations/)
  - V4 recently released, might be a good time to give it a try
  - Based on XML inputs, apparently extensible with C# (not sure about this part)
  - [Real-life example with a lot of documentation](https://helgeklein.com/blog/real-world-example-wix-msi-application-installer/)
  - [Examples that run custom code](https://cs.github.com/?scopeName=All+repos&scope=&q=WixQuietExec64+path%3A*.wxs)
- [Inno Setup](https://jrsoftware.org/isinfo.php)
  - Extensible with Pascal (might be a bit saner than NSIS' but still...)
  - Does not depend on MSI technology
  - Another veteran with good track record
  - [Good CLI options](https://jrsoftware.org/ishelp/index.php?topic=setupcmdline)
  - Example: [this is what Julia uses](https://github.com/JuliaLang/julia/blob/8b4b1f03149904d399d1887ffe3c2b3b60375c9c/contrib/windows/build-installer.iss)
  - [Plenty of cases where installers run custom code](https://cs.github.com/?scopeName=All+repos&scope=&q=path%3A*.iss+%22ShellExecute%22)

I think both WiX and InnoSetup might be capable replacements, but none feels like  The One ✨  either. Both have good CLI options. WiX uses MSI under the hood, which might be better for corporate users. InnoSetup seems closer to NSIS in spirit, with Pascal Scripting.

So, to sum up, no obvious answer.

**Comment by jaimergp** [comment]
On the CLI-only side, I can see some tools (`scoop`) are using combinations like:

- Powershell [`irm`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-restmethod?view=powershell-7.3) fetches a [script](https://raw.githubusercontent.com/scoopinstaller/install/master/install.ps1) and pipes it to [`iex`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-expression?view=powershell-7.3) (e.g. `irm get.scoop.sh | iex`). 
  - This is like `wget script.sh | sh` in Unix world. 
  - Not that we need to pipe, but we could follow `scoop`'s setup for a CLI-only installer.
  - We would need to reimplement many of the features.
  - No uninstaller though, but that's on #642

**Comment by RahulARanger** [comment]
Hi @jaimergp, I have a bit of experience with Inno setup. 

It would be possible to cover the cases with the Inno setup. most of them lie in the fact that we can write the scripts with pascal, or if required we can take control of the required things using PowerShell.

I have worked on a [Project](https://github.com/RahulARanger/CO-PO-Mapping) It was packaged using Innosetup.

If possible, can I try to contribute to this?

**Comment by jaimergp** [comment]
Hi @RahulARanger, that's very exciting!

I'd say we approach this with some caution though, because it would add some maintenance burden to a team that's already stretched a bit too thin. I'd like to know what @conda/constructor thinks about it.

Ideally, if you ask me, I would envision something like `constructor` offering some kind of plugin mechanism so users can use other "installer backends", and then you could maintain your own plugin facilitating that. Given enough time and if it gets a lot of traction, such a plugin could be "absorbed" in `constructor` itself. But of course, that plugin system does not exist right now so I might be dreaming a bit too much :D

**Comment by RahulARanger** [comment]
True, it is hard to build a plugin system and even expect the same in the other platforms.

**Comment by jaimergp** [comment]
An intermediate step for our own sanity is to rebuild the NSI templates with some Jinja, [like `pynsist` is doing](https://github.com/takluyver/pynsist/blob/master/nsist/pyapp.nsi). Quite cleaner!

**Comment by jaimergp** [comment]
Not stale
